Pairing logic fix

aanh1009 opened the pull request:
## 🚀 What does this PR do?

> A short summary of what’s changed and why.
pulled the frontend-backend restructuring from main, built API flows for the pairing logic on the backend, integrate the APIs to execute pairing on the frontend
<img width="154" alt="Screenshot 2025-05-17 102514" src="https://github.com/user-attachments/assets/ba19704f-1478-442c-a29e-7d90e0e1a01f" />
<img width="158" alt="Screenshot 2025-05-17 102453" src="https://github.com/user-attachments/assets/895f771c-ed19-41cd-8db7-05cc87f36a29" />
<img width="722" alt="image" src="https://github.com/user-attachments/assets/23c4a3dd-c82a-40f9-b77d-a2ed7e2ace10" />

---

## ✅ Checklist

- [ ] Code compiles & runs
- [ ] Lint checks pass
- [ ] Updated `.env.example` if needed
- [ ] PR linked to related issue/ticket (if applicable)

---

## 🔍 How to test this

> Describe how reviewers can test this — steps, screenshots, etc.

---

## 🧠 Extra Notes

> Any decisions made? Tradeoffs? Gotchas?

Quinxie51 committed:
create pairing now page and add it to on boarding

aanh1009 committed:
separated onboarding and pairing, added pairing logic

aanh1009 committed:
pulled refactoring from main

aanh1009 committed:
update tsconfig.json

aanh1009 committed:
added pairing logic, still needs to debug the room deleting api

aanh1009 committed:
completed pairing logic

aanh1009 committed:
completed pairing logic

aanh1009 committed:
changed packages

tranguv commented on backend/src/controllers/roomController.ts:
```diff
@@ -0,0 +1,55 @@
+//ROLE: validate user input
+
+//Handle user registration
+export async function createRoom(req: any, res: any) {
```
Cast type any to every controllers
import { Request, Response } from 'express';

interface SomeParam {}
export async function createRoom(req: Request<SomeParam>, res:Response)

ThanhDatVu111 replied:
Done

tranguv commented on backend/src/controllers/roomController.ts:
```diff
@@ -0,0 +1,55 @@
+    if (!room_id) {
+      return res.status(400).json({ error: 'Missing required room_id parameter' });
+    }
+
```
Should we check if room exists

aanh1009 replied:
> Should we check if room exists

I fixed this

tranguv commented on backend/src/services/roomService.ts:
```diff
@@ -0,0 +1,24 @@
+  });
+}
+
+export async function joinRoom(input: any) {
```
Should we check if a user is already in a room, check if room exists and is available?

aanh1009 replied:
> Should we check if a user is already in a room, check if room exists and is available?

the current code already checks this

tranguv commented on backend/src/services/roomService.ts:
```diff
@@ -0,0 +1,24 @@
+
+export async function registerRoom(input: any) {
+  //Receive “raw” data from the controller and pass it to the model
+  return roomModel.createRoom({
```
Can a user create multiple rooms hmm? And what if they already joined a room

ThanhDatVu111 replied:
user cannot join multiple rooms

ThanhDatVu111 replied:
if they already joined a room, the only option to join another room is to get out of that room.

tranguv commented on backend/src/services/roomService.ts:
```diff
@@ -0,0 +1,24 @@
+}
+
+export async function deleteRoom(input: any) {
+  return roomModel.deleteRoom({
```
Should we check if the roomId exists?

aanh1009 replied:
> Should we check if the roomId exists?

i fixed this

tranguv commented on frontend/src/app/(onboard)/join-room.tsx:
```diff
@@ -0,0 +1,125 @@
+        setPartnerCode={setPartnerCode}
+        onFinish={connectRoom}
+        error={error}
+      ></PairingStep>
```
/>

tranguv commented on backend/src/controllers/roomController.ts:
```diff
@@ -0,0 +1,55 @@
+      return res.status(400).json({ error: 'Missing required fields' });
+    }
+    //pass the request body to the userService
+    await roomService.joinRoom(req.body);
```
I was wondering what this will return, what if it return an error, should we handle error here?

aanh1009 replied:
> I was wondering what this will return, what if it return an error, should we handle error here?

this does not return anything. it it succeeds, it sets the status to 204 (which means successful but no response), and there is already error handling below

tranguv reviewed (dismissed):
From my understanding, your `models` folder look like a `repositories` for me (Database logic only, for data-in/data-out and no business logic) so you can just probably rename it for more clarification; services will be handle business logic. You can create another folder to define your database schema (models). For your reference: https://mingyang-li.medium.com/production-grade-node-js-typescript-folder-structure-for-2024-f975edeabefd

Quinxie51 committed:
fix UI of pairing room

aanh1009 committed:
tested UI + updated packages

aanh1009 committed:
handled invalid code/no code bugs for join-room

aanh1009 committed:
handled invalid code/no code bugs for join-room

aanh1009 committed:
handle duplicated room creation

aanh1009 committed:
pull from main and resolved merge conflicts

yourvivian commented on backend/src/models/roomModel.ts:
```diff
@@ -0,0 +1,48 @@
+    .from('room')
+    .update({
+      user_2: attrs.user_2,
+      filled: true,
+    })
```
How about checking if `user_2` is null or not to determine if the room is "filled" instead of a dedicated column?

ThanhDatVu111 replied:
Done

yourvivian commented on backend/src/routes/roomRoutes.ts:
```diff
@@ -0,0 +1,12 @@
+router.put('/joinRoom', joinRoom);
+router.get('/checkRoom', checkRoom);
+router.delete('/deleteRoom/:room_id', deleteRoom);
+
```
"It is not correct to put the verbs in REST URIs. REST uses nouns to represent resources, and HTTP methods (GET, POST, PUT, DELETE, etc.) are then used to perform actions on those resources, effectively acting as verbs." ([ref](https://restfulapi.net/resource-naming/#:~:text=It%20is%20not%20correct%20to%20put%20the%20verbs%20in%20REST%20URIs.%20REST%20uses%20nouns%20to%20represent%20resources%2C%20and%20HTTP%20methods%20(GET%2C%20POST%2C%20PUT%2C%20DELETE%2C%20etc.)%20are%20then%20used%20to%20perform%20actions%20on%20those%20resources%2C%20effectively%20acting%20as%20verbs.))

if we see "/room" and method is `PUT`, we know we're updating the room

ThanhDatVu111 replied:
done

yourvivian commented on backend/src/routes/roomRoutes.ts:
```diff
@@ -0,0 +1,12 @@
+import { Router } from 'express';
+import { checkRoom, createRoom, joinRoom, deleteRoom } from '../controllers/roomController';
+import { checkPrime } from 'crypto';
+
```
remove unused import

ThanhDatVu111 replied:
Done

yourvivian commented on backend/src/controllers/roomController.ts:
```diff
@@ -0,0 +1,82 @@
+    return res.json({ response });
+
+    return response;
+  } catch (err: any) {
```
unreachable code

ThanhDatVu111 replied:
Deleted it!

yourvivian commented on backend/src/models/roomModel.ts:
```diff
@@ -0,0 +1,48 @@
+export async function deleteRoom(attrs: { room_id: string }) {
+  const { error } = await supabase.from('room').delete().eq('room_id', attrs.room_id);
+  if (error) throw error;
+}
```
Should return a status or the deleted row here for confirmation

ThanhDatVu111 replied:
Done

yourvivian commented on backend/src/controllers/roomController.ts:
```diff
@@ -0,0 +1,82 @@
+      .json({ data: room }); // ← send back JSON payload
+  } catch (err: any) {
+    if (err.code === '23505') {
+      return res.status(409).json({ error: 'room already created' });
```
Let's create a file to define PostgreSQL error codes as an enum or constants since we're likely to reuse them

ThanhDatVu111 replied:
Done

yourvivian commented on frontend/src/apis/room.ts:
```diff
@@ -0,0 +1,87 @@
+
+export interface deleteRoomRequest {
+  room_id: string;
+}
```
TypeScript interfaces should be PascalCase

yourvivian replied:
We should move the interfaces to a shared types file, such as `room.ts` in a src/type folder to keep the API logic and type definitions separate

ThanhDatVu111 replied:
Done

yourvivian commented on frontend/src/apis/room.ts:
```diff
@@ -0,0 +1,87 @@
+
+  } catch (err: any) {
+    throw err;
+  }
```
The `catch (err: any)` block only catches Ts runtime errors (like network errors, JSON parsing errors, etc.), not HTTP errors. You would need to check `response.ok` and throw an err

ThanhDatVu111 replied:
Done

yourvivian commented on frontend/src/app/(onboard)/onboardingFlow.tsx:
```diff
@@ -1,11 +1,13 @@
 import DateTimePicker from '@react-native-community/datetimepicker';
 import * as ImagePicker from 'expo-image-picker';
 import Button from '@/components/Button';
+import * as Clipboard from 'expo-clipboard';
```
removed unused import

ThanhDatVu111 replied:
Done

yourvivian commented on frontend/src/app/(onboard)/onboardingFlow.tsx:
```diff
@@ -1,11 +1,13 @@
 import { useRouter } from 'expo-router';
-import { View, Text, Platform, Image, TextInput } from 'react-native';
+import { View, Text, Platform, Image, TextInput, Alert } from 'react-native';
 import DateTimePicker from '@react-native-community/datetimepicker';
```
removed unused import

ThanhDatVu111 replied:
Done

yourvivian commented on backend/src/server.ts:
```diff
@@ -3,6 +3,8 @@ import cors from 'cors';
 import UserRouter from './routes/userRoutes';
+import RoomRouter from './routes/roomRoutes';
+import router from './routes/userRoutes';
 // import other routers like TripRouter, NotificationRouter if needed
```
remove unused import

ThanhDatVu111 replied:
Done

yourvivian commented on frontend/src/app/(onboard)/join-room.tsx:
```diff
@@ -0,0 +1,137 @@
+}
+
+const joinRoom = () => {
+  const router = useRouter();
```
React components should be PascalCase

ThanhDatVu111 replied:
Done

yourvivian commented on frontend/src/app/(onboard)/join-room.tsx:
```diff
@@ -0,0 +1,137 @@
+  const [partnerCode, setPartnerCode] = useState('');
+  const [error, setError] = useState('');
+
+  const connectRoom = async () => {
```
Should we add a loading state to prevent double submissions and show progress during API calls? (disable the button while the request is in progress)

ThanhDatVu111 replied:
Done

yourvivian commented on frontend/src/app/(onboard)/onboardingFlow.tsx:
```diff
@@ -183,23 +185,39 @@ const OnboardingFlow = () => {
   const { userId, email } = useLocalSearchParams();
   const router = useRouter();
+  const roomId = crypto.randomUUID();
 
```
`crypto.randomUUID()` is called on every render. How do we make sure the same roomId is used throughout the onboarding flow?

yourvivian reviewed (commented):
Good work!

Some comments:
- Avoid using `any`. Define and use TypeScript interfaces for input params to catch errors early and improve code clarity
- use prettier on all files you worked on to format code
- Some commented-out lines are redundant because the code is already self-explanatory. You can read: https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/

yourvivian reviewed (commented):
Good work!

Some comments:
- Avoid using `any`. Define and use TypeScript interfaces for input params to catch errors early and improve code clarity
- use prettier on all files you worked on to format code
- Some commented-out lines are redundant because the code is already self-explanatory. You can read: https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/

ThanhDatVu111 committed:
Resolve remaining conflicts

ThanhDatVu111 committed:
refactor(frontend): Improve onboarding flow, room pairing, and entry guard logic

### Summary
- Refactored onboarding flow to streamline user steps (name, birthdate, profile picture).
- Updated  for better room pairing functionality and UI improvements.
- Integrated  API into entry guard () to ensure proper navigation:
  - Redirects to pairing if the room is not filled.
  - Redirects to onboarding or home based on user and room data.
- Organized type definitions into a dedicated  folder for better maintainability.
- Updated API endpoints and improved error handling for room and user management.

ThanhDatVu111 committed:
feat(backend): Add fetchRoom API and refactor room routes

### Summary
- Added  API () to retrieve the  status of a room by .
- Refactored  to improve error handling and validation.
- Introduced  for handling Supabase-specific errors.
- Updated  for RESTful structure and added the new route.
- Improved type definitions for room-related requests and responses.

Quinxie51 committed:
fix UI for paring code

Quinxie51 committed:
make the pairing code more significant

ThanhDatVu111 committed:
feat: Add loading state to useEntryGuard and separate verification logic

- Added a  state to  to display a loading indicator during async operations.
- Improved navigation flow by showing a spinner and Loading... text while fetching user and room data.
- Handled errors gracefully by stopping the loading state when an error occurs.
- Enhanced user experience with smoother page transitions during redirects.
- Separated email verification logic into its own  file for better modularity and maintainability.

ThanhDatVu111 reviewed (approved):
lgtm for now!

ThanhDatVu111 committed:
minor change in roomModel

ThanhDatVu111 committed:
delete migration

MC2114 reviewed (approved):
Breaking down the pairing logic into createRoom, checkRoom, fetchRoom, joinRoom functions made the code very readable! When tested on my machine, everything works and the comments detail the purpose of the code very thoroughly. Great work!
